### PR TITLE
[Merged by Bors] - chore: Remove nonterminal `simp at`

### PR DESCRIPTION
**Pull request by BoltonBailey**
Removes nonterminal uses of `simp at`. Replaces most of these with instances of `simp? ... says`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by digama0** [comment]
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

File: Mathlib/Algebra/FreeAlgebra.lean [hunk 1]
```diff
@@ -461,7 +461,8 @@ noncomputable def equivMonoidAlgebraFreeMonoid :
       · simp
         rfl
       · intro x y ih
-        simp at ih
+        simp only [MonoidAlgebra.of_apply, AlgHom.coe_comp, Function.comp_apply,
+          MonoidAlgebra.lift_single, one_smul, AlgHom.coe_id, id_eq] at ih
         simp [ih])
```
Comment (YaelDillies) [inline Mathlib/Algebra/FreeAlgebra.lean:466]: This simp was already terminal. It's only followed by another `simp`.

File: Mathlib/Analysis/SpecialFunctions/PolarCoord.lean [hunk 2]
```diff
@@ -42,7 +42,7 @@ def polarCoord : LocalHomeomorph (ℝ × ℝ) (ℝ × ℝ) where
     rcases eq_or_ne θ 0 with (rfl | h'θ)
     · simpa using hr
     · right
-      simp at hr
+      simp only [mem_Ioi] at hr
       simpa only [ne_of_gt hr, Ne.def, mem_setOf_eq, mul_eq_zero, false_or_iff,
         sin_eq_zero_iff_of_lt_of_lt hθ.1 hθ.2] using h'θ
```
Comment (YaelDillies) [inline Mathlib/Analysis/SpecialFunctions/PolarCoord.lean:47]: Same for this one

File: Mathlib/CategoryTheory/Limits/Shapes/Biproducts.lean [hunk 3]
```diff
@@ -714,7 +714,7 @@ instance {ι} (f : ι → Type*) (g : (i : ι) → (f i) → C)
           · simp at h
             by_cases w : j = j'
             · cases w
-              simp at h
+              simp only [heq_eq_eq, forall_true_left] at h
```
Comment (YaelDillies) [inline Mathlib/CategoryTheory/Limits/Shapes/Biproducts.lean:717]: And here too

File: Mathlib/CategoryTheory/Monad/Algebra.lean [hunk 4]
```diff
@@ -204,7 +204,7 @@ def adj : T.free ⊣ T.forget :=
       homEquiv_naturality_right := by
         intros
         -- This doesn't look good:
-        simp
+        simp only [forget_obj, free_obj_A, forget_map]
```
Comment (YaelDillies) [inline Mathlib/CategoryTheory/Monad/Algebra.lean:207]: Yet another one

**Review by YaelDillies — COMMENTED** [review]
I'm not sure what the value of such a PR is. On one hand, you make the proofs less likely to break due to the addition of a lemma in the simp set. On the other hand, proofs now refer explicitly to more lemmas that are susceptible to change in ways that a fancy tactic call wouldn't have noticed.

**Comment by BoltonBailey** [comment]
> I'm not sure what the value of such a PR is. On one hand, you make the proofs less likely to break due to the addition of a lemma in the simp set. On the other hand, proofs now refer explicitly to more lemmas that are susceptible to change in ways that a fancy tactic call wouldn't have noticed.

I am sympathetic to this argument. However, I think most maintainers are on the side of "nonterminal simps are bad". I have gotten reviews before telling me to remove nonterminal simps. The point of this PR is to make way for #7496 so that reviewers hopefully don't have to do this as often.

I will revert the things you suggest and also make the style PR more lax so that they don't conflict.

File: Archive/Arithcc.lean [hunk 5]
```diff
@@ -321,7 +321,8 @@ theorem compiler_correctness :
   -- 5.III
   case sum =>
     rename_i e_s₁ e_s₂ e_ih_s₁ e_ih_s₂
-    simp
+    simp only [compile._eq_3, List.append_assoc, List.singleton_append, List.cons_append,
```
Comment (semorrison) [inline Archive/Arithcc.lean:324]: I don't think we should be referring to `._eq_3` manually.

**Comment by BoltonBailey** [comment]
Note that the zulip thread on this PR suggested we should be using `simp says` for this kind of thing.

File: Archive/Arithcc.lean (hunk 5 above)
Comment (BoltonBailey) [reply Archive/Arithcc.lean]: ok I have reverted all the things that look like this.

**Comment by BoltonBailey** [comment]
Not sure what this error means. Annoyingly, my file builds just fine in the IDE, but CI is saying it's broken.
```
Error: ./././Mathlib/Algebra/FreeAlgebra.lean:577:2: error: Failed to parse tactic output: simp only [AlgHom.ext_iff, AlgHom.coe_id, id_eq, AlgHom.coe_comp, Subalgebra.coe_val, Function.comp_apply] at
of_id 
```

**Comment by semorrison** [comment]
You should be able to reproduce the additional testing that `says` does in CI using `lake build -KCI Mathlib.X.Y.Z`. Or put `set_option says.verify true` in the failing files.

**Comment by semorrison** [comment]
I merged master, and then looked at the first two errors: both were resolved by replacing the current RHS with what `says` is now suggesting.

I'll hand it back over to you to look at the remainder?

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**